### PR TITLE
Create text file editor, add possiblity to display/edit/save script files

### PR DESCRIPTION
**Pull request by ptomin**
изображение

### CODE REVIEW

**Comment by ptomin** [comment]
Add (*) to Text Editor tab title if file was changed
изображение

**Comment by ptomin** [comment]
Add possibility to navigate by stack frames at diagnostics tab.
изображение
User can click at stack frame and navigate to corresponding line of file.

File: src/Gui/TextFileViewerServiceImpl.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        {
+        }
+
+        public void DisplayFile(string fileName, int? line)
+        {
+            var pane = new TextFileEditorInteractor(fileName);
+            var windowType = typeof(TextFileEditorInteractor).Name;
+            var title = Path.GetFileName(fileName);
+            var frame = ShowWindow(windowType, title, fileName, pane);
+            pane = ((TextFileEditorInteractor) frame.Pane);
```
Comment (uxmal) [inline src/Gui/TextFileViewerServiceImpl.cs:43]: Why is this necessary? Shouldn't be sufficient to just call `pane.DisplayFile()` below without resetting the  `pane` variable?

File: src/Scripts/Python/PythonModule.cs [hunk 2]
```diff
@@ -65,10 +66,12 @@ public override void FireEvent(ScriptEvent @event, Program program)
...
             {
-                eventListener.Error(
-                    new NullCodeLocation(Filename),
+                var scriptError = CreateError(
+                    Filename,
                     ex,
-                    "An error occurred while running the Python script.");
+                    "An error occurred while running the Python script.",
+                    engine);
+                eventListener.Error(scriptError);
```
Comment (uxmal) [inline src/Scripts/Python/PythonModule.cs:74]: Again, I'd prefer an extension method for this.

File: src/UserInterfaces/WindowsForms/DecompilerShellUiService.cs [hunk 3]
```diff
@@ -105,6 +105,7 @@ public IWindowFrame CreateWindow(string windowType, string windowTitle, IWindowP
             framesByTab.Add(tabPage, frame);
             this.form.DocumentTabs.TabPages.Add(tabPage);
             this.form.DocumentTabs.SelectedTab = tabPage;
+            pane.Frame = frame;
```
Comment (uxmal) [inline src/UserInterfaces/WindowsForms/DecompilerShellUiService.cs:108]: Nice catch!

File: src/UserInterfaces/WindowsForms/WindowsDecompilerEventListener.cs [hunk 4]
```diff
@@ -269,6 +270,35 @@ public void Error(ICodeLocation location, Exception ex, string message, params o
             diagnosticSvc.Error(location, ex, message, args);
         }
 
+        public void Error(ScriptError scriptError)
```
Comment (uxmal) [inline src/UserInterfaces/WindowsForms/WindowsDecompilerEventListener.cs:273]: Again, this code, specific to `Reko.Core.Scripts` should be in an extension method.

File: src/Gui/Controls/TextFileEditorInteractor.cs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        private ITextFileEditor? editor;
+        private readonly string fileName;
+        private bool textLoaded;
+        private bool textChanged;
+
+        public TextFileEditorInteractor(string fileName)
+        {
+            this.fileName = fileName;
+            this.textLoaded = false;
+            this.textChanged = false;
```
[the remark this answers is not in the document]
Comment (ptomin) [reply src/Gui/Controls/TextFileEditorInteractor.cs]: Done in a732f9a

File: src/Gui/TextFileViewerServiceImpl.cs (hunk 1 above)
Comment (ptomin) [reply src/Gui/TextFileViewerServiceImpl.cs:43]: if window already exist `ShowWindow` ignores `pane` argument. We need to use `frame.Pane` to get access to correct pane (both new or existing one)

File: src/Gui/Controls/TextFileEditorInteractor.cs [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+            {
+                using var rdr = fsSvc.CreateStreamReader(
+                    fileName, Encoding.UTF8);
+                editor.TextBox.Text = rdr.ReadToEnd();
+                this.textLoaded = true;
+                this.textChanged = false;
+            }
+            catch (Exception ex)
+            {
+                uiSvc.ShowError(ex, $"Couldn't open file '{fileName}'");
```
[the remark this answers is not in the document]
Comment (ptomin) [reply src/Gui/Controls/TextFileEditorInteractor.cs]: Done in 40beb6d

File: src/Gui/Design/ScriptFileDesigner.cs [hunk 7]
```diff
@@ -26,13 +29,21 @@ namespace Reko.Gui.Design
 {
     public class ScriptFileDesigner : TreeNodeDesigner
     {
+        private ScriptFile? scriptFile;
```
[the remark this answers is not in the document]
Comment (ptomin) [reply src/Gui/Design/ScriptFileDesigner.cs]: Done in df54038

**Comment by ptomin** [comment]
Show stack trace at separate dialog
изображение

**Comment by ptomin** [comment]
Highlight stack frame line at text editor
изображение
